libnavigation-router/src/main/java/com/mapbox/navigation/route/internal/RouterWrapper.kt

```diff
@@ -0,0 +1,227 @@
+                        }
+                    }
+                }
+            )
+        }
+    }
+
+    override fun getRouteRefresh(
+        route: DirectionsRoute,
+        legIndex: Int,
+        callback: RouteRefreshCallback
+    ): Long {
+        val routeOptions = route.routeOptions()
+        val requestUuid = route.requestUuid()
+        val routeIndex = route.routeIndex()?.toIntOrNull()
+        if (routeOptions == null || requestUuid == null || routeIndex == null) {
+            val errorMessage =
+                """
+                   Route refresh failed because of a null param:
+                   routeOptions = $routeOptions
+                   requestUuid = $requestUuid
+                   routeIndex = $routeIndex
+                """.trimIndent()
+
+            logW(TAG, Message(errorMessage))
+
+            callback.onError(
+                RouteRefreshError("Route refresh failed", Exception(errorMessage))
+            )
+
+            return REQUEST_FAILURE
+        }
+
+        val userAndProfile = "${routeOptions.user()}/${routeOptions.profile()}"
+        val refreshOptions = RouteRefreshOptions(
+            requestUuid,
+            routeIndex,
+            legIndex,
+            RoutingProfile(userAndProfile),
+        )
+
+        return router.getRouteRefresh(refreshOptions, route.toJson()) { result, _ ->
+            result.fold(
+                {
+                    mainJobControl.scope.launch {
+                        val errorMessage =
+                            """
+                           Route refresh failed.
+                           error = ${it.error}
+                           legIndex = $legIndex
+                           code = ${it.code}
+                           requestId = ${it.requestId}
+                            """.trimIndent()
+
+                        logW(TAG, Message(errorMessage))
+
+                        callback.onError(
+                            RouteRefreshError("Route refresh failed", Exception(errorMessage))
+                        )
+                    }
+                },
+                {
+                    mainJobControl.scope.launch {
+                        val refreshedRoute =
+                            withContext(ThreadController.IODispatcher) {
+                                DirectionsRoute.fromJson(
+                                    it,
+                                    routeOptions,
+                                    route.requestUuid()
+                                )
+                            }
+                        callback.onRefresh(refreshedRoute)
+                    }
+                }
+            )
+        }
+    }
+
+    override fun cancelRouteRequest(requestId: Long) {
+        router.cancelRequest(requestId)
+    }
+
+    override fun cancelRouteRefreshRequest(requestId: Long) {
+        router.cancelRequest(requestId)
+    }
```

Comment:
@Guardiola31337 to do that we need to change `interface Router` API. can we make breaking changes now?

Reply:
This is a great question. I'd vote yes, it's a good opportunity to narrow down the number of APIs exposed, we're still in early `rc` stage and this change is directly related to OP, making the overall user experience better and more friendly. What do you think @mapbox/navigation-android?

Reply:
I'd vote no - the `Router` interface exposes a separate `getRoute` and `getRouteRefresh` methods so it makes sense that it also exposes a `cancelRouteRequest` and `cancelRouteRefreshRequest` separately. Just because our native implementation uses the same method for both actions under thehood shouldn't drive the design for the interface itself.